### PR TITLE
Update Gradle Plugin to 7.4.2 + Fix Overflowing Poetry

### PR DESCRIPTION
**Pull request by davissuber**
This pull request does the following:
* Update Android Gradle Plugin to 7.4.2 to stay someone consistent with other open source libraries like RIBs
    * Specify Java language to version 11 for Gradle
    * Miscellaneous build script fixes
* Attempt to address https://github.com/uber/motif/issues/255 by tidying up parameter type resolution - we were previously generating unresolved generic types first and then fixing them post generation - however, this 2-step process causes a stack overflow error if the unresolved generic type is self-referencing - this pull request merge the 2-step parameter resolution process into one to avoid the crash.

### CODE REVIEW

File: compiler/src/main/kotlin/motif/compiler/XFunSpec.kt [hunk 1]
```diff
@@ -50,23 +47,13 @@ object XFunSpec {
...
 
-    val builder = overriding(methodElement)
-    builder.returns(javaToKotlinType(returnType))
-
-    var i = 0
-    val size = builder.parameters.size
-    val resolvedParameterTypes = method.parameterTypes
-    while (i < size) {
-      val parameter = builder.parameters[i]
-      val type = javaToKotlinType(resolvedParameterTypes[i])
```
Comment (davissuber) [inline compiler/src/main/kotlin/motif/compiler/XFunSpec.kt:61]: This is the real `type` we want, but we can only get it from `method` (which has the enclosing generic context), but not `methodElement` (which does not). Hence, we pass `method.parameterTypes` into the `overriding` function.

File: compiler/src/main/kotlin/motif/compiler/XFunSpec.kt [hunk 2]
```diff
@@ -90,9 +77,10 @@ object XFunSpec {
                 .forEach { funBuilder.addTypeVariable(it) }
     */
 
-    method.parameters.forEach {
+    method.parameters.forEachIndexed { index, parameter ->
       funBuilder.addParameter(
-          ParameterSpec.builder(it.name, it.type.typeName.toKTypeName()).build())
```
Comment (davissuber) [inline compiler/src/main/kotlin/motif/compiler/XFunSpec.kt:95]: This `toKTypeName` eventually creates an infinite loop somewhere in kotlin poet due to the unresolved type references itself in the generic type parameter.

File: tests/src/main/java/testcases/KT006_reference_self/Child.kt [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package testcases.KT006_reference_self
+
+@motif.Scope
+interface Child {
+
+    @motif.Objects
+    abstract class Objects : ObjectComponent<Bar>
```
Comment (davissuber) [inline tests/src/main/java/testcases/KT006_reference_self/Child.kt:22]: This is the key ingredient of the repro case, that an `Objects` inherits a function that contains a self-referencing generic parameter

File: samples/sample-kotlin-ksp/build.gradle [hunk 4]
```diff
@@ -26,8 +30,8 @@ android {
     }
 
     compileOptions {
-        sourceCompatibility JavaVersion.VERSION_1_8
-        targetCompatibility JavaVersion.VERSION_1_8
+        sourceCompatibility JavaVersion.VERSION_11
```
Comment (davissuber) [inline samples/sample-kotlin-ksp/build.gradle:33]: Not sure how useful these are, much of the code is already in kotlin.

File: gradle/dependencies.gradle [hunk 5]
```diff
@@ -42,13 +42,13 @@ ext.deps = [
...
                 gradlePlugins: [
-                        android: 'com.android.tools.build:gradle:4.2.0',
+                        android: 'com.android.tools.build:gradle:7.4.2',
                         kotlin: "org.jetbrains.kotlin:kotlin-gradle-plugin:${versions.kotlin}",
                         ksp: "com.google.devtools.ksp:com.google.devtools.ksp.gradle.plugin:${versions.ksp}",
                         dokka: "org.jetbrains.dokka:dokka-gradle-plugin:${versions.dokka}",
                         mavenPublish: 'com.vanniktech:gradle-maven-publish-plugin:0.18.0',
                         spotless: "com.diffplug.spotless:spotless-plugin-gradle:5.11.0",
-                        shadow: "com.github.jengelman.gradle.plugins:shadow:6.1.0",
+                        shadow: "com.github.johnrengelman:shadow:8.1.0",
```
Comment (davissuber) [inline gradle/dependencies.gradle:51]: This looks questionable - but [it is still the same library](https://imperceptiblethoughts.com/shadow/changes/#v8-1-0-2023-02-26) - upgrade is needed for the new gradle plugin version.

**Comment by davissuber** [comment]
I have tested [88c9c0a](https://github.com/uber/motif/pull/257/commits/88c9c0a30f8809ff85efc6b1de2a47cb52980034) by successfully building uber's 3 big apps with it.

**Comment by muandrew** [comment]
Thanks for checking internally also.
Note for future changes/contributors the changes to the build files can be done in another PR that can be quickly accepted while we get some time to review changes to the logic.

**Pull request merged**